Mesh_3/include/CGAL/Mesh_3/tet_soup_to_c3t3.h

```diff
@@ -0,0 +1,447 @@
+                                     c->vertex((i+3)%4),
+                                     c->vertex((i+2)%4));
+
+      // set the infinite_vertex's incident cell
+      if(tr.infinite_vertex()->cell() == Cell_handle())
+        tr.infinite_vertex()->set_cell(opp_c);
+
+      // add the facets to the incident cells map
+
+      // the only finite facet
+      it->second.push_back(std::make_pair(opp_c, inf_vert_position_in_opp_c));
+
+      add_infinite_facets_to_incident_cells_map<Tr>(opp_c, inf_vert_position_in_opp_c,
+                                                    incident_cells_map);
+    }
+  }
+}
+
+template<class Tr>
+bool assign_neighbors(Tr& tr,
+                      const std::map<std::set<typename Tr::Vertex_handle>,
+                      std::vector<std::pair<typename Tr::Cell_handle,
+                      int> > >& incident_cells_map)
+{
+  typedef typename Tr::Vertex_handle                              Vertex_handle;
+  typedef typename Tr::Cell_handle                                Cell_handle;
+  typedef std::set<Vertex_handle>                                 Facet;
+  typedef std::pair<Cell_handle, int>                             Incident_cell;
+  typedef std::map<Facet, std::vector<Incident_cell> >            Incident_cells_map;
+
+  // 4 facets per cell, each facet shared by 2 cells
+  if(incident_cells_map.size() != tr.number_of_cells() * 2)
+    return false;
+
+  typename Incident_cells_map::const_iterator icit = incident_cells_map.begin();
+  for(; icit!=incident_cells_map.end(); ++icit)
+  {
+    const std::vector<Incident_cell>& adjacent_cells = icit->second;
+    if(adjacent_cells.size() != 2)
+      return false;
+
+    Cell_handle c0 = adjacent_cells[0].first;
+    int i0 = adjacent_cells[0].second;
+    Cell_handle c1 = adjacent_cells[1].first;
+    int i1 = adjacent_cells[1].second;
+
+    tr.tds().set_adjacency(c0, i0, c1, i1);
+  }
+  return true;
+}
+
+template<class Tr, bool c3t3_loader_failed>
+bool build_triangulation(Tr& tr,
+                         const std::vector<typename Tr::Point>& points,
+                         const std::vector<boost::array<int,5> >& finite_cells,
+                         const std::vector<boost::array<int,4> >& infinite_cells,
+                         const std::map<boost::array<int,3>, int>& border_facets)
+{
+  typedef typename Tr::Vertex_handle            Vertex_handle;
+  typedef typename Tr::Cell_handle              Cell_handle;
+  typedef std::set<Vertex_handle>               Facet;
+
+  // associate to a face the two (at most) incident tets and the id of the face in the cell
+  typedef std::pair<Cell_handle, int>                   Incident_cell;
+  typedef std::map<Facet, std::vector<Incident_cell> >  Incident_cells_map;
+
+  Incident_cells_map incident_cells_map;
+  std::vector<Vertex_handle> vertex_handle_vector(points.size() + 1); // id to vertex_handle
+
+  CGAL_precondition(!points.empty() && !finite_cells.empty());
```

Comment:
I find it strange that the treatment for empty finite cells differ in the two  functions.